src/GHCMod.hs

```diff
@@ -650,20 +653,22 @@ legacyInteractiveCmd = withParseCmd [] go
        return ghcModiVersion
    go _ = throw $ InvalidCommandLine (Left "legacy-interactive")
 
-checkAction :: ([t] -> a) -> [t] -> a
+checkAction :: (Functor m, GmEnv m) => ([FilePath] -> m a) -> [FilePath] -> m a
 checkAction _ []         = throw $ InvalidCommandLine (Right "No files given.")
-checkAction action files = action files
+checkAction action files = mapM gmeFileCwd files >>= action
 
-locAction :: String -> (String -> Int -> Int -> a) -> [String] -> a
-locAction _ action [file,_,line,col] = action file (read line) (read col)
-locAction _ action [file,  line,col] = action file (read line) (read col)
+locAction :: (Functor m, GmEnv m) => String -> (String -> Int -> Int -> m a) -> [String] -> m a
+locAction _ action [file,_,line,col] = gmeFileCwd file >>= \f -> action f (read line) (read col)
+locAction _ action [file,  line,col] = gmeFileCwd file >>= \f -> action f (read line) (read col)
 locAction cmd _ _ = throw $ InvalidCommandLine (Left cmd)
 
-locAction' :: String -> (String -> Int -> Int -> Expression -> a) -> [String] -> a
-locAction' _ action [f,_,line,col,expr] = action f (read line) (read col) (Expression expr)
-locAction' _ action [f,  line,col,expr] = action f (read line) (read col) (Expression expr)
+locAction' :: (Functor m, GmEnv m) => String -> (String -> Int -> Int -> Expression -> m a) -> [String] -> m a
+locAction' _ action [f,_,line,col,expr] = gmeFileCwd f >>= \f' -> action f' (read line) (read col) (Expression expr)
+locAction' _ action [f,  line,col,expr] = gmeFileCwd f >>= \f' -> action f' (read line) (read col) (Expression expr)
 locAction' cmd _ _ = throw $ InvalidCommandLine (Left cmd)
 
+gmeFileCwd :: (Functor m, GmEnv m) => FilePath -> m FilePath
+gmeFileCwd f = liftM (</> f) $ optCwd . gmOptions <$> gmeAsk
```

Comment:
For most cases it won't, since it affects only user-supplied argument

```
$ ghc-mod check `pwd`/test/test-pandoc-crossref.hs 
test/test-pandoc-crossref.hs:8:1:Warning: The import of ‘Data.Monoid’ is redundant  except perhaps to import instances from ‘Data.Monoid’To import instances alone, use: import Data.Monoid()
```

Lint, however, is a weird outlier in this case (which probably needs to be fixed, btw):

```
$ ghc-mod lint `pwd`/test/test-pandoc-crossref.hs 
/home/livid/github/pandoc-crossref/test/test-pandoc-crossref.hs:28:1: Error: Eta reduceFound:  go x = id xWhy not:  go = id
/home/livid/github/pandoc-crossref/test/test-pandoc-crossref.hs:28:8: Error: EvaluateFound:  id xWhy not:  x
$ ghc-mod lint test/test-pandoc-crossref.hs 
test/test-pandoc-crossref.hs:28:1: Error: Eta reduceFound:  go x = id xWhy not:  go = id
test/test-pandoc-crossref.hs:28:8: Error: EvaluateFound:  id xWhy not:  x
```